### DISCUSSION TITLE
[Python API] Reusing connection with islice() limit

### DISCUSSION THREAD

**Discussion by alex-mirkin** [erezsh/reladiff#35] [Q&A] [answered]
When using python api to compare a list of tables, I'm making the connection outside the loop and then reusing it for all the tables I need to compare.
But after getting the first x `islice`ed rows from diffs, the differ continues to check the rest of the table and doesn't stop until the whole table is compared or the connection is closed.
This is problematic for huge tables or tables with many diffs.

Is this an expected behavior?
Should I create and close a connection for each table?

Simplified version of the code:
```
# create db connections
mysql_conn = connect(mysql_conn_details, thread_count=MAX_THREADS)
snowflake_conn = connect(snowflake_conn_details)

tables = ['foo.table1', 'bar.table2']

# iterate over tables to compare
for table in tables:
    # create a connection to the table
    table_seg_mysql = TableSegment(database, table_path, key_columns, extra_columns)
    table_seg_snowflake = TableSegment(database, table_path, key_columns, extra_columns)

    diffs = diff_tables(table_seg_mysql, table_seg_snowflake)

    # limit the number of diffs to check
    iterator = islice(diffs, 10)

    # check the diffs
    for sign, row in iterator:
        print(sign, row)

# close connections
mysql_conn.close()
snowflake_conn.close()
```

**Comment by erezsh** [comment]
I think the scheduling is supposed to be lazy with a lookahead. If it is eager, then maybe it's a bug. I will look into it.

**Reply by erezsh** [reply]
@alex-mirkin I was wrong, the algorithm is currently eager, and will keep diffing the entire tables in the background.

I was able to change its behavior to be lazy, it's a simple fix. However, even when lazy, the thread-pool will remain alive on idle, and apparently it even survives GC (there is no "daemon" equivalent for gc shutdown), so it needs to be explicitly shut-down.

I might be able to fix it if I write my own thread-queue, instead of using the python built-in.

But I think the simpler solution is to just provide a `close()` method that will shutdown the current diff (without affecting the db connections).

Something like:

```python
    diffs = diff_tables(table_seg_mysql, table_seg_snowflake)

    # limit the number of diffs to check
    iterator = islice(diffs, 10)

    # check the diffs
    for sign, row in iterator:
        print(sign, row)

    diffs.close()
```

Or maybe even use `with`:

```python
with diff_tables(table_seg_mysql, table_seg_snowflake) as diff:
    ...
```

What do you think?

**Reply by erezsh** [reply]
@alex-mirkin I created a PR with an attempted fix: https://github.com/erezsh/reladiff/pull/38

Let me know if it works! (after calling .close() ofc)

**Reply by alex-mirkin** [reply]
Shutting down the current diff without affecting the db connections sounds like an elegant solution for this.
I'll test and update.
Thanks!

**Reply by alex-mirkin** [reply]
I've tested and the `diffs.close()` works as expected.
Looks good!

Adding the `with` support will be a nice bonus.

**Reply by erezsh** [reply]
Added support for `with` in PR #39.

**Reply by alex-mirkin** [reply]
I'm seeing a strange behavior after the [lazy commit](https://github.com/erezsh/reladiff/commit/0afbe65da265f05dbd5c4dbcc086e636b58705a4) in a specific scenario:
When comparing two tables in a loop and the first table is really huge one - after the diff finished all the tables, the program will not exit - because of stuck threads.
<img width="381" alt="image" src="https://github.com/user-attachments/assets/594bfbd8-e4c9-422f-bb6a-fc6826cc9af7">

The number of stuck threads = `max_threadpool_size`

Anything I can check to see why it happens?

**Reply by erezsh** [reply, marked as answer]
Thanks for letting me know. Most likely it is caused by this commit: 0afbe65da265f05dbd5c4dbcc086e636b58705a4 

I tried to fix it in this PR: https://github.com/erezsh/reladiff/pull/40

Could you give it a try and let me know if it solved the problem for you?

If not, I will revert the entire commit, and that should fix it for sure (if I'm not wrong).

**Reply by alex-mirkin** [reply]
I can confirm that PR: https://github.com/erezsh/reladiff/pull/40 fixes the issue for this test case.
Thanks!
